common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java

```diff
@@ -222,7 +223,7 @@ public void sendMergedBlockMetaReq(
     handler.addRpcRequest(requestId, callback);
     RpcChannelListener listener = new RpcChannelListener(requestId, callback);
     channel.writeAndFlush(
-      new MergedBlockMetaRequest(requestId, appId, shuffleId, reduceId)).addListener(listener);
+      new MergedBlockMetaRequest(requestId, appId, shuffleId, shuffleSequenceId, reduceId)).addListener(listener);
```

Comment:
Why do we need to modify the fetch side requests? When will this request for shuffle data of an older shuffleSequenceId?

Reply:
Other than for validation, offhand I cant seem to see a good usecase for this.
Do you see concerns with this change @otterc ?

Reply:
My concern is that when the shuffleSequenceId is part of the message, then it seems the reader will also read the shuffle data for an older `shuffleSequenceId`.  If this is never the case, then it makes it confusing that this field exist in the message.

Reply:
Do you think a better name or adding a comment here would help? Isn't it similar like the mapId set in the fetch message for regular shuffle?

Reply:
It is not an issue with the name of this field or just a matter of adding a comment. You are proposing to add a field to these messages suggesting that these messages will be used to fetch shuffle data of older `shuffleSequenceId` which is never the case. So if the reducers are always fetching latest `shuffleSequenceId` corresponding to a shuffleId then the server can keep just some metadata information and figure this out itself. As mentioned offline this metadata will be less than few kbs. 

> Isn't it similar like the mapId set in the fetch message for regular shuffle?

I don't see any similarities with this and the mapId in the regular fetch message.

Reply:
`mapId` is similar to the `shuffleSequenceId` because even there we don't fetch shuffle data for an older `mapId` mapped to the same `mapIndex`.

I can't understand how introducing this parameter to the message is suggestive that older `shuffleSequenceId` can also be fetched for a shuffle ID. Semantics of the usage of a variable can always be clearly documented. I am not convinced that adding this variable to the fetch protocol message is confusing, if it is really confusing it can always be well documented.

Reply:
> mapId is similar to the shuffleSequenceId because even there we don't fetch shuffle data for an older mapId mapped to the same mapIndex.

If the fetcher doesn't include a mapId in the request, how will the server know which shuffle block to serve? 

> I can't understand how introducing this parameter to the message is suggestive that older shuffleSequenceId can also be fetched for a shuffle ID. Semantics of the usage of a variable can always be clearly documented. I am not convinced that adding this variable to the fetch protocol message is confusing, if it is really confusing it can always be well documented.

If you are adding a field to a message that implies it has a specific need. However, in this case there is no need to include that field.  On the other hand, tracking this information on the server is straightforward with minimal memory usage.

Reply:
> If the fetcher doesn't include a mapId in the request, how will the server know which shuffle block to serve?

Agree, it is not exactly same.

As mentioned in other comment, in terms of correctness both the approaches should work fine. Here it is just a choice of implementation. We can agree to disagree. I still prefer passing `shuffleSequenceId` in the fetch protocol as it feels simpler and straightforward to me. If there are other strong reasons for not doing so other than the ones mentioned earlier, I would definitely consider them and change the implementation accordingly.

Reply:
> I still prefer passing shuffleSequenceId in the fetch protocol as it feels simpler and straightforward to me. If there are other strong reasons for not doing so other than the ones mentioned earlier, 

As I mentioned in my earlier [comment](https://github.com/apache/spark/pull/33034#discussion_r657623667), IMO these are strong enough reason to not modify the protocol 
- Adding a field that is not really needed. This change trickles down to the various other classes and APIs like `BlockStoreClient`, `ExternalBlockStoreClient`, `OneForOneBlocker`, `ExternalBlockHandler`, etc.  
Instead the alternative is just a self-contained simple change to track this metadata in `RemoteBlockPushResolver`.

- Adding this field in the message suggests that the fetch request could be for older `shuffleSequenceId`.  You said we can add a clarifying comment, but if there is no need to make such a change then what's the point of adding a comment to explain all this.

Reply:
What about the following scenario:
1. An indeterminate stage generates the shuffle data for a given shuffle.
2. Downstream reduce stage experienced shuffle fetch failure, leading to retry of the indeterminate stage.
3. Tasks from the retry of the indeterminate stage start pushing blocks, which would lead to invalidating the shuffle data from the 1st attempt.
4. In the meantime, we might still have dangling tasks from the first failed reduce stage trying to fetch shuffle blocks corresponding to the 1st attempt of the indeterminate stage.

Is the above scenario possible with indeterminate stage retry, and would we run into issues if the seq ID is only used on the push side but not the fetch side?

Reply:
When an indeterminate stage is retried due to fetch failure, spark will fail all children stages which depend on that stage across all jobs and fail all of them so they are retried. See [here](https://github.com/apache/spark/blob/b5a15035851bfba12ef1c68d10103cec42cbac0c/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1807).

Reply:
Although Spark always cancels tasks from the failed stage, there could be a race condition between canceling and task running, right? So, it seems possible to me to fetch the older shuffle data.

Reply:
Even if a task continues to fetch older shuffle data, it wont be allowed to 'commit' the results since the stage it ended up fetching for would have been cancelled (both for ResultStage and for ShuffleMapStage, the result of the task will end up being ignored).
This is equivalent to current state as well, if task does not honour thread interruption and continues to fetch data/process.

Reply:
> Even if a task continues to fetch older shuffle data, it wont be allowed to 'commit' the results since the stage it ended up fetching for would have been cancelled (both for ResultStage and for ShuffleMapStage, the result of the task will end up being ignored).

That's true.

I just want to confirm that there could be a case for the task to fetch older shuffle data. 

(The thread is too long. Sorry if I miss some context which already mentioned before. )

Reply:
I did have a theoretical concern regarding pushing blocks actually - not reading (that, as discussed above, should be fine) : and I had discussed it with @otterc and @zhouyejoe 

The scenario is as follows: Suppose a stage is cancelled/failed and a reattempt starts - with one or more tasks of the failed stage still running and pushing data. If ESS "knows" of a new shuffle sequence id (due to some other task from new stage attempt pushing data), it can reject all pushes for the 'older' sequence ids.

The current code flow does not prevent the push of old stage task from completing - but also does not cause any incorrect shuffle reads in child stages (as earlier stage attempt has failed, it will never 'finalize' that shuffle output - and so never added to MapOutputTracker as merged location).
But we were debating on whether the added validation while processing push blocks is worth it (this will require propagating the shuffle seq id to those requests as well).

Thoughts @Ngone51 ?

Reply:
Not sure what's "validation" you actually mean here. Could you explain more?

IIUC, while pushing blocks, we have to add seq id for blocks no matter the ESS would reject or not.

Reply:
@mridulm Are you saying we might not need to pass shuffleSequenceId to finalizeShuffleMerge call? Whenever the finalize call happens it is always for the latest shuffleSequenceId?

Reply:
Discussed with @venkata91 on this issue again. One thing I think that we didn't all realize before is the implication of introducing this shuffle sequence ID. Before introducing this sequence ID, a block push request is identifying the partition it needs to be merged into uniquely with <appID, shuffleID, partitionID>. Similarly, the shuffle service is uniquely identifying the location of the latest merged shuffle file for a given partition using <appID, shuffleID, partitionID>. So the information passed in the block push request is sufficient for the shuffle service to identify the latest merged shuffle file for that shuffle partition on its node.

Now with shuffle sequence ID, that changes the things. A block push request (for an indeterminate shuffle map stage) is using <appID, shuffleID, shuffleSeqID, partitionID> to uniquely identify the merged shuffle file that it needs to be merged into. However, the shuffleSeqID passed in the block push request might not be the latest shuffleSeqID tracked by the shuffle service for the latest merged shuffle file for the given shuffle partition. The seqID in the block push request could be smaller than the latest seqID tracked by ESS (if the task is a dangling task from a failed previous attempt) or larger (if the task is from a newer attempt of the indeterminate shuffle map stage). When the latest shuffleSeqID tracked by ESS is still present in memory, there isn't much of a problem. However, when this in-memory metadata is cleaned after merge finalization, we would encounter issues.

In scenario where a dangling task from a failed 1st stage attempt is still pushing blocks after the 2nd stage attempt gets merge finalized, or 1st stage attempt gets merge finalized but downstream reduce stage experiences shuffle fetch failures leading to retry of the stage and tasks from 2nd stage attempt pushing blocks, the shuffleSeqID passed in the block push request would be either smaller or larger than the shuffleSeqID corresponding to the latest merged shuffle file maintained on the ESS node. Since the metadata about what's the latest shuffleSeqID for a previously merge finalized shuffle is already cleaned due to merge finalization, we would have the following options:
1) We maintain a separate data structure to keep track of the latest shuffleSeqID encountered so far for a given shuffle ID for each ESS, and this information is not deleted during merge finalization. The concern with this approach is that, we would need to access this separate data structure during both handling of block push and merge finalization, and this could lead to the same concurrency issue we saw earlier with handling multiple app attempts.
2) We rely on the merged shuffle files to tell what's the latest shuffleSeqID for a given shuffle ID encountered by the ESS so far. This would require searching the local files for a given appID, shuffleID, partition ID with various shuffle sequence ID. However, there does not seem to be a clear upper bound for when to stop this search. If the block push request has a sequence ID of 2, it is possible that the ESS does not hold merged shuffle files for sequence ID 1, 2, 3 but holds merged shuffle files for sequence ID 4.
3) We do not care what's the latest sequence ID encountered by a given ESS. Instead we keep all shuffle files for all sequence IDs encountered for a shuffle ID on the ESS. This would then require the fetch protocol to specify sequence ID as well, as proposed in this PR.

@otterc @mridulm @Ngone51 thoughts?

Reply:
Sorry, I was thinking of fetch side - and wrote push side ... my bad.
You are right, on push/write side we do need it for sure - my entire comment was about read side of things.
Let me edit my comment about to make it clear - sorry for the confusion !

Reply:
To update, my current thought is to add the sequence id for reads as well (**only** for merged blocks) and validate/fail read request is from task with an older sequence id. While not required for current state of codebase, it will help as things evolve.
Thoughts ?

Reply:
It does seem that not having `shuffleSequenceId` is causing confusion and requires a lot of explanation. So, I am fine with adding this new field in just the merged fetch protocols.

Also discussed with @Victsm offline about his comments and we have a way to address some of the open issues with the implementation. @Victsm or @venkata91 will elaborate about the approach.

Reply:
> Since the metadata about what's the latest shuffleSeqID for a previously merge finalized shuffle is already cleaned due to merge finalization, we would have the following options....

I prefer option 3. And we can clean up merged shuffle files at the time of

https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala#L173-L181 

whether clean it at executor side directly or send a message to ESS to ask clean. (I currently not sure which way is better.)

And I agree to add the seq id at the fetching side as well (although an older fetch should never happen), which seems more consistent with the pushing side to me.

Reply:
After having another discussion with @Victsm, now it is clear on how we want to proceed. Following are the changes that needs to be added:
1. Pass `shuffleSequenceId` as part of push, fetch and finalize protocol - seems like we are all in agreement with this change now.
2. Keep track of `shuffleId -> shuffleSequenceId (latest attempt)` mapping as part of the existing Map we have by adding an additional level for `shuffleSequenceId` -  `Map<ShuffleId, Map<ShuffleSequenceId, Map<PartitionId, AppShufflePartitionInfo>>>` With this throughout the lifecycle of the application, we will keep track of the latest `shuffleId -> shuffleSequenceId` mapping, this way we can reject the push/fetch requests made for a older shuffleSequenceId for a shuffleId. Also as part of the finalization request for an `<appid, shuffleId, shuffleSequenceId>, we can simply close all the files tracked for the `shuffleId, shuffleSequenceId` and remove this metadata but still keep the `shuffleSequenceId` key around. This is required for rejecting invalid push/fetch requests.
3. Whenever a push request is received for a `<shuffleId, shuffleSequenceId>` which is greater than the current one tracked as part of the metadata, close/clean up all the files for the older attempt in an async manner. This way the stale attempts shuffle data can be gradually cleaned.

Given that we are all now convinced on the protocol changes required, I'm planning to separate out the protocol changes and the implementation detail into separate PR so that we can get the protocol changes PR merged before 3.2 branch cut. Thoughts @mridulm @Ngone51 @Victsm @otterc ?

Reply:
> With this throughout the lifecycle of the application, we will keep track of the latest shuffleId -> shuffleSequenceId mapping, this way we can reject the push/fetch requests made for a older shuffleSequenceId for a shuffled. 

To confirm, this's actually a variant of option 1, right?

Reply:
@ngone51 This is taking key aspects of both approach 1 and 3. We are keeping the metadata of `shuffleId -> shuffleSequenceId` in ESS for both rejecting stale push/fetch requests as well as for clean up of old attempts shuffle data.

Reply:
Got it.

Reply:
@venkata91 The branch-3.2 is already cut. I think we can just proceed with this PR. The idea sounds good to me. Could you update accordingly? We have very limited time to merge it into branch-3.2.

Reply:
@ngone51 This PR depends on some of the changes made by @zhouyejoe as part of the other PR handling multiple app attempts as we want to add another level to the Map for shuffleSequenceId in addition to the appAttempt